nce/NCE-Pairwise-SM/train.py

```diff
@@ -0,0 +1,420 @@
+                        new_train_pos["ext_feat"].append(ext_feat_i)
+
+                        near_answer = question2answer[qid_i]["neg"][near_id]["answer"]
+                        if near_answer.size()[0] > max_len_q:
+                            max_len_q = question_i.size()[0]
+                        if near_answer.size()[0] > max_len_a:
+                            max_len_a = near_answer.size()[0]
+                        ext_feat_neg = question2answer[qid_i]["neg"][near_id]["ext_feat"]
+                        new_train_neg["answer"].append(near_answer)
+                        new_train_neg["question"].append(question_i)
+                        new_train_neg["ext_feat"].append(ext_feat_neg)
+
+            elif label_i == 1:
+
+                if aid_i not in question2answer[qid_i]["neg"]:
+                    answer_i = answer_i[answer_i != 1]
+                    question2answer[qid_i]["neg"][aid_i] = {"answer": answer_i}
+
+                question2answer[qid_i]["neg"][aid_i]["feature"] = features[i].data.cpu().numpy()
+                question2answer[qid_i]["neg"][aid_i]["ext_feat"] = ext_feat_i
+
+                if epoch == 1:
+                    if qid_i not in q2neg:
+                        q2neg[qid_i] = []
+
+                    q2neg[qid_i].append(aid_i)
+
+        # pack the selected pos and neg samples into the torchtext batch and train
+        if epoch != 1:
+            true_batch_size = len(new_train_neg["answer"])
+            if true_batch_size != 0:
+                for j in range(true_batch_size):
+                    new_train_neg["answer"][j] = F.pad(new_train_neg["answer"][j],
+                                                       (0, max_len_a - new_train_neg["answer"][j].size()[0]), value=1)
+                    new_train_pos["answer"][j] = F.pad(new_train_pos["answer"][j],
+                                                       (0, max_len_a - new_train_pos["answer"][j].size()[0]), value=1)
+                    new_train_pos["question"][j] = F.pad(new_train_pos["question"][j],
+                                                       (0, max_len_q - new_train_pos["question"][j].size()[0]), value=1)
+                    new_train_neg["question"][j] = F.pad(new_train_neg["question"][j],
+                                                       (0, max_len_q - new_train_neg["question"][j].size()[0]), value=1)
+
+                pos_batch = get_batch(new_train_pos["question"], new_train_pos["answer"], new_train_pos["ext_feat"],
+                                      true_batch_size)
+                neg_batch = get_batch(new_train_neg["question"], new_train_neg["answer"], new_train_neg["ext_feat"],
+                                      true_batch_size)
+
+                optimizer.zero_grad()
+                output = pw_model([pos_batch, neg_batch])
+
+                '''
+                debug code
+                '''
+                cmp = output[:, 0] <= output[:, 1]
```

Comment:
For the debug and evaluation code, could you add a class for them. So we can evaluate the loss, evaluation metrics easily for different sets.